server/services/userAdminService.js

```diff
@@ -111,49 +111,92 @@ module.exports = function createUserService(nomisClientBuilder, userClient) {
     }
   }
 
-  async function getIncompleteRoUsers(token) {
-    const incomplete = await userClient.getIncompleteRoUsers()
-    if (isEmpty(incomplete)) {
+  async function getIncompleteRoUsers(userName) {
+    const systemTokens = await signInService.getClientCredentialsTokens(userName)
+
+    const bookingIds = await userClient.getCasesRequiringRo()
+    if (isEmpty(bookingIds)) {
+      return []
+    }
+
+    const all = await findRequiredRos(systemTokens.token, bookingIds)
+    if (isEmpty(all)) {
       return []
     }
 
-    const bookingIds = incomplete.map(u => u.bookingId)
-    const offenders = await getOffenderNomis(token, bookingIds)
+    const required = uniqueByAssignedId(all)
+    const incomplete = await findIncomplete(required)
+    if (isEmpty(incomplete)) {
+      return []
+    }
 
-    return addMappingDetails(incomplete, offenders)
+    return addOffenderNomis(systemTokens.token, incomplete)
   }
 
-  async function getOffenderNomis(token, bookingIds) {
-    try {
-      const nomisClient = nomisClientBuilder(token)
-      const offenders = await nomisClient.getOffenderSentencesByBookingId(bookingIds, false)
+  async function findRequiredRos(token, bookingIds) {
+    const required = await Promise.all(
+      bookingIds.map(async bookingId => {
+        try {
+          const { com } = await prisonerService.getCom(bookingId, token)
+          if (com) {
+            return { assignedId: com.deliusId, assignedName: com.name, bookingId }
+          }
+        } catch (error) {
+          logger.warn(`Error in getCom for incomplete users, booking id ${bookingId}`, error.stack)
+        }
+      })
+    )
 
-      if (isEmpty(offenders)) {
-        return null
-      }
+    return required.filter(e => e)
```

Comment:
not sure what you are filtering out here is `e` a boolean?

Reply:
Can be anything - just removes where null/undefined

Reply:
```
return required.filter(Boolean)
```